// angular-cesium/src/services/cesium/cesium.service.js
import { Injectable, NgZone, Optional } from '@angular/core';
import { ViewerFactory } from '../viewer-factory/viewer-factory.service';
import { ViewerConfiguration } from '../viewer-configuration/viewer-configuration.service';
export var CesiumService = (function () {
    function CesiumService(ngZone, viewerFactory, viewConf) {
        this.ngZone = ngZone;
        this.viewerFactory = viewerFactory;
        this.viewConf = viewConf;
    }
    CesiumService.prototype.init = function (mapContainer) {
        var _this = this;
        this.ngZone.runOutsideAngular(function () {
            var options = _this.viewConf ? _this.viewConf.viewerOptions : undefined;
            _this.cesiumViewer = _this.viewerFactory.createViewer(mapContainer, options);
        });
    };
    CesiumService.prototype.getViewer = function () {
        return this.cesiumViewer;
    };
    CesiumService.prototype.getScene = function () {
        return this.cesiumViewer.scene;
    };
    CesiumService.prototype.getMinimumZoom = function () {
        return this.getScene().screenSpaceCameraController.minimumZoomDistance;
    };
    CesiumService.prototype.setMinimumZoom = function (amount) {
        this.getScene().screenSpaceCameraController.minimumZoomDistance = amount;
    };
    CesiumService.prototype.getMaximumZoom = function () {
        return this.getScene().screenSpaceCameraController.maximumZoomDistance;
    };
    CesiumService.prototype.setMaximumZoom = function (amount) {
        this.getScene().screenSpaceCameraController.maximumZoomDistance = amount;
    };
    CesiumService.prototype.setEnableTilt = function (isTilt) {
        this.getScene().screenSpaceCameraController.enableTilt = isTilt;
    };
    CesiumService.prototype.morphTo2D = function (duration) {
        if (duration === void 0) { duration = 2.0; }
        this.getScene().morphTo2D(duration);
    };
    CesiumService.prototype.morphTo3D = function (duration) {
        if (duration === void 0) { duration = 2.0; }
        this.getScene().morphTo3D(duration);
    };
    CesiumService.prototype.morphToColumbusView = function (duration) {
        if (duration === void 0) { duration = 2.0; }
        this.getScene().morphToColumbusView(duration);
    };
    CesiumService.prototype.flyTo = function (options) {
        this.getViewer().camera.flyTo(options);
    };
    CesiumService.decorators = [
        { type: Injectable },
    ];
    CesiumService.ctorParameters = function () { return [
        { type: NgZone, },
        { type: ViewerFactory, },
        { type: ViewerConfiguration, decorators: [{ type: Optional },] },
    ]; };
    return CesiumService;
}());
