// ts-mockito/lib/MethodStubCollection.js
"use strict";
var MethodStubCollection = (function () {
    function MethodStubCollection() {
        this.hadMoreThanOneBehavior = false;
        this.items = [];
    }
    MethodStubCollection.prototype.add = function (item) {
        this.items.push(item);
        if (this.items.length > 1) {
            this.hadMoreThanOneBehavior = true;
        }
    };
    MethodStubCollection.prototype.getFirstMatchingAndRemove = function (args) {
        var index = this.getFirstMatchingIndex(args);
        var result = this.getFirstMatching(args);
        if (index > -1) {
            this.items.splice(index, 1);
        }
        return result;
    };
    MethodStubCollection.prototype.getFirstMatching = function (args) {
        for (var _i = 0, _a = this.items; _i < _a.length; _i++) {
            var item = _a[_i];
            if (item.isApplicable(args)) {
                return item;
            }
        }
        return null;
    };
    MethodStubCollection.prototype.getHadMoreThanOneBehavior = function () {
        return this.hadMoreThanOneBehavior;
    };
    MethodStubCollection.prototype.hasMatching = function (args) {
        return this.getFirstMatchingIndex(args) > -1;
    };
    MethodStubCollection.prototype.getFirstMatchingIndex = function (args) {
        var index = 0;
        for (var _i = 0, _a = this.items; _i < _a.length; _i++) {
            var item = _a[_i];
            if (item.isApplicable(args)) {
                return index;
            }
            index++;
        }
        return -1;
    };
    return MethodStubCollection;
}());
exports.MethodStubCollection = MethodStubCollection;
