// ts-mockito/lib/mockito-node-tester.js
"use strict";
var ts_mockito_1 = require("./ts-mockito");
var Generic = (function () {
    function Generic() {
    }
    return Generic;
}());
exports.Generic = Generic;
var Temp = (function () {
    function Temp() {
        this.sampleVar = "test";
    }
    Temp.prototype.sampleMethod = function (param) {
        this.sampleVar = param.toString();
        return this.sampleVar;
    };
    Temp.prototype.sampleMethod2 = function (first, second) {
        return 10;
    };
    return Temp;
}());
exports.Temp = Temp;
function main() {
    var myMock = ts_mockito_1.mock(Temp);
    ts_mockito_1.when(myMock.sampleMethod(1)).thenReturn("jeden");
    ts_mockito_1.when(myMock.sampleMethod(2)).thenReturn("dwa");
    ts_mockito_1.when(myMock.sampleMethod(3)).thenReturn("trzy");
    ts_mockito_1.when(myMock.sampleMethod2("sampleString", 3)).thenReturn(1);
    console.log(">>", ts_mockito_1.instance(myMock).sampleMethod(1));
    console.log(">>", ts_mockito_1.instance(myMock).sampleMethod(2));
    console.log(">>", ts_mockito_1.instance(myMock).sampleMethod(2));
    ts_mockito_1.verify(myMock.sampleMethod(ts_mockito_1.anyNumber())).thrice();
}
exports.main = main;
